### DISCUSSION TITLE
Lambda layers shouldn't be sorted at synthesis

### DISCUSSION THREAD

**Discussion by ctgardner** [aws/aws-cdk#26395] [General]
> Lambda extracts the layers in the order (low to high) listed by the function. Lambda merges folders with the same name, so if the same file appears in multiple layers, the function uses the version in the last extracted layer.
>
> Source: https://docs.aws.amazon.com/lambda/latest/dg/invocation-layers.html#invocation-layers-accessing

Lambda layers are supposed to be extracted (and merged) in the order they're registered. This is the expected and current behaviour exhibited by AWS CDK (and CloudFormation), by default.

However, when `@aws-cdk/aws-lambda:recognizeLayerVersion` is enabled, the layers are sorted at synthesis. Potentially, this reorders layers and causes breaking changes to Lambda functions which rely upon the documented extract-and-merge behaviour.

https://github.com/aws/aws-cdk/blob/f19ecefcdde712dfd951106bec3b1f850b66f2a8/packages/%40aws-cdk/aws-lambda/lib/function.ts#L1058-L1060

According to a test included with this change in https://github.com/aws/aws-cdk/pull/20150 (https://github.com/aws/aws-cdk/commit/f19ecefcdde712dfd951106bec3b1f850b66f2a8), a reason for sorting layers is to ensure the same hash is calculated for Lambda functions which register the same layers in a _different_ order, but are otherwise identical.

https://github.com/aws/aws-cdk/blob/f19ecefcdde712dfd951106bec3b1f850b66f2a8/packages/%40aws-cdk/aws-lambda/test/function-hash.test.ts#L210-L230

But because layers can overwrite each other, the order in which they're extracted is crucial and Lambda functions which register the same layers in a _different_ order should always have a different hash.

Regardless of the previous point, why is it necessary to _mutate_ the layers array to calculate the Lambda function hash? A sorted _copy_ can be used to calculate the same hash. This would cause the layers to maintain the original ordering and be extracted in the expected order.

In summary, Lambda layers can overwrite each other when extracted, so the order in which they're extracted is crucial. Lambda functions which list the same layers in a different order aren't necessarily identical and shouldn't have the same hash. Per the documentation, layers should be extracted in the order listed by the function. Extracting Lambda layers in any other order is unexpected and is a potentially breaking change.

**Comment by pasujemito** [comment]
Is there an update on this?

**Reply by raypappa** [reply]
No. because fuck users.

https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-lambda/lib/function.ts#L1519-L1521
